### PR TITLE
Allow chunks to use the persistent data API

### PR DESCRIPTION
**Pull request by brodybeckwith**

### CODE REVIEW

File: Spigot-Server-Patches/0420-Allow-chunks-to-use-the-persistent-data-API.patch [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+ import org.bukkit.entity.Entity;
++import org.bukkit.persistence.PersistentDataContainer;
+ import org.bukkit.plugin.Plugin;
+ 
+ public class CraftChunk implements Chunk {
+@@ -39,6 +42,8 @@ public class CraftChunk implements Chunk {
+     private final int z;
+     private static final DataPaletteBlock<IBlockData> emptyBlockIDs = new ChunkSection(0).getBlocks();
+     private static final byte[] emptyLight = new byte[2048];
++    private static final CraftPersistentDataTypeRegistry DATA_TYPE_REGISTRY = new CraftPersistentDataTypeRegistry();
```
Comment (Black-Hole) [inline Spigot-Server-Patches/0420-Allow-chunks-to-use-the-persistent-data-API.patch]: // Paper comments missing

**Review by aikar — DISMISSED** [review]
This has a lot of work to actually be complete. This is effectively what my https://github.com/PaperMC/Paper/pull/1326 was but using the existing bukkit api that came later.

<details>
<summary>Ignore, incorrect information</summary>


You must handle data conversions. When a chunk is data converted, you must then have Data Fixers registered that takes the chunks current and then new version, then find every sub data point that can also be migrated, and pass in the source and target versions to it too.

Namely item stacks is what I know will be needed. Not sure what else this api supports storing.

</details>
edit: see below

**Comment by BrodyBeckwith** [comment]
> 
> 
> This has a lot of work to actually be complete. This is effectively what my #1326 was but using the existing bukkit api that came later.
> 
> You must handle data conversions. When a chunk is data converted, you must then have Data Fixers registered that takes the chunks current and then new version, then find every sub data point that can also be migrated, and pass in the source and target versions to it too.
> 
> Namely item stacks is what I know will be needed. Not sure what else this api supports storing.

Are you saying that the data needs to be converted on say version updates (1.14.4 -> 1.15)? If tha'ts the case whats stopping entities or itemstacks from having the same problem, which are essentially identical to what I did to chunks.

**Comment by aikar** [comment]
It has came to my attention this API from Spigot actually doesn't support directly storing complex data, and only supports storing primitives, which effectively gives a huge screw you to handling version updates...
so nevermind on that.

File: Spigot-Server-Patches/0420-Allow-chunks-to-use-the-persistent-data-API.patch (hunk 1 above)
Comment (brodybeckwith) [reply Spigot-Server-Patches/0420-Allow-chunks-to-use-the-persistent-data-API.patch]: Should be fixed.

**Comment by brodybeckwith** [comment]
I'm going to go ahead and close this and add some additional functionality for saving persistent data for block locations as well and submit a PR in likely a couple days.

**Pull request closed without merging**